// @reportei/n8n-nodes-reportei/dist/nodes/Reportei/ReporteiTrigger.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReporteiTrigger = void 0;
class ReporteiTrigger {
    constructor() {
        this.description = {
            displayName: 'Reportei Trigger',
            name: 'reporteiTrigger',
            icon: 'file:reportei.svg',
            group: ['trigger'],
            version: 1,
            description: 'Handle Reportei events via webhooks',
            defaults: {
                name: 'Reportei Trigger',
            },
            inputs: [],
            outputs: ['main'],
            credentials: [
                {
                    name: 'reporteiApi',
                    required: true,
                },
            ],
            properties: [
                {
                    displayName: 'Project Name or ID',
                    name: 'clientId',
                    type: 'options',
                    default: '',
                    required: true,
                    description: 'Select the project to which you want to add the trigger. Choose from the list, or specify an ID using an expression. Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code-examples/expressions/">expression</a>.',
                    typeOptions: {
                        loadOptionsMethod: 'getClients',
                    },
                },
                {
                    displayName: 'Event Type',
                    name: 'event',
                    type: 'options',
                    options: [
                        { name: 'Automation Executed', value: 'automation_executed' },
                        { name: 'Control Goal Met', value: 'control_goal_met' },
                        { name: 'Control Goal Not Met', value: 'control_goal_not_met' },
                        { name: 'Dashboard Created', value: 'dashboard_created' },
                        { name: 'Report Created', value: 'report_created' },
                        { name: 'Timeline Milestone Added', value: 'timeline_milestone_added' },
                    ],
                    default: 'report_created',
                    required: true,
                    description: 'Which event to subscribe to',
                },
                {
                    displayName: 'Custom Webhook Base URL',
                    name: 'customWebhookBaseUrl',
                    type: 'string',
                    default: '',
                    placeholder: 'https://your-domain.com',
                    description: 'If you are behind a tunnel or need a specific domain, provide your public base URL here. The node will replace the default localhost address with this URL.',
                },
            ],
            webhooks: [
                {
                    name: 'default',
                    httpMethod: 'POST',
                    responseMode: 'onReceived',
                    path: 'webhook',
                },
            ],
        };
        this.webhookMethods = {
            default: {
                async checkExists() {
                    return false;
                },
                async create() {
                    const clientId = this.getNodeParameter('clientId');
                    const event = this.getNodeParameter('event');
                    let webhookUrl = this.getNodeWebhookUrl('default');
                    const customBase = this.getNodeParameter('customWebhookBaseUrl', 0) || '';
                    if (customBase) {
                        webhookUrl = replaceBaseUrl(webhookUrl, customBase);
                    }
                    const body = {
                        client_id: clientId,
                        source: 'n8n',
                        url: webhookUrl,
                        event_type: event,
                    };
                    const response = await this.helpers.httpRequestWithAuthentication.call(this, 'reporteiApi', {
                        method: 'POST',
                        url: 'https://app.reportei.com/api/v1/webhooks/subscribe',
                        body,
                        json: true,
                    });
                    if (response === null || response === void 0 ? void 0 : response.id) {
                        this.getWorkflowStaticData('node').webhookId = response.id;
                    }
                    return true;
                },
                async delete() {
                    const clientId = this.getNodeParameter('clientId');
                    const event = this.getNodeParameter('event');
                    let webhookUrl = this.getNodeWebhookUrl('default');
                    const customBase = this.getNodeParameter('customWebhookBaseUrl', 0) || '';
                    if (customBase) {
                        webhookUrl = replaceBaseUrl(webhookUrl, customBase);
                    }
                    const body = {
                        client_id: clientId,
                        source: 'n8n',
                        event_type: event,
                        url: webhookUrl,
                    };
                    await this.helpers.httpRequestWithAuthentication.call(this, 'reporteiApi', {
                        method: 'POST',
                        url: 'https://app.reportei.com/api/v1/webhooks/unsubscribe',
                        body,
                        json: true,
                    });
                    delete this.getWorkflowStaticData('node').webhookId;
                    return true;
                },
            },
        };
        this.methods = {
            loadOptions: {
                async getClients() {
                    const returnData = [];
                    const response = await this.helpers.httpRequestWithAuthentication.call(this, 'reporteiApi', {
                        method: 'GET',
                        url: 'https://app.reportei.com/api/v1/clients?per_page=1000',
                        json: true,
                    });
                    for (const client of response.data) {
                        returnData.push({
                            name: client.name,
                            value: client.id,
                        });
                    }
                    return returnData;
                },
            },
        };
    }
    async webhook() {
        const req = this.getRequestObject();
        const bodyData = req.body;
        return {
            workflowData: [
                [
                    {
                        json: bodyData,
                    },
                ],
            ],
        };
    }
}
exports.ReporteiTrigger = ReporteiTrigger;
function replaceBaseUrl(originalUrl, customBase) {
    const path = extractPathFromUrl(originalUrl);
    const sanitizedBase = customBase.replace(/\/+$/, '');
    return `${sanitizedBase}${path}`;
}
function extractPathFromUrl(fullUrl) {
    try {
        const urlObj = new URL(fullUrl);
        return urlObj.pathname + (urlObj.search || '');
    }
    catch {
        return '/webhook';
    }
}
