// @reportei/n8n-nodes-reportei/dist/credentials/ReporteiApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReporteiApi = void 0;
class ReporteiApi {
    constructor() {
        this.name = 'reporteiApi';
        this.displayName = 'Reportei API';
        this.documentationUrl = 'https://app.reportei.com/docs/api';
        this.properties = [
            {
                displayName: 'API Token',
                name: 'apiToken',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
                description: 'Enter your Reportei API token here',
                required: true,
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    ContentType: 'application/json',
                    Authorization: '=Bearer {{$credentials.apiToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://app.reportei.com/api/v1',
                url: '/me',
                method: 'GET',
            },
        };
    }
}
exports.ReporteiApi = ReporteiApi;
